File: frontend/src/api/ai.js
// ============================================================================
// api/ai.js
// ============================================================================
// Talks to backend/routes/ai.py. All Claude interactions go through here.
// ============================================================================

import client from './client.js'

export async function chat(message, history = [], datasetId = null) {
  const { data } = await client.post('/api/ai/chat', {
    message,
    history,
    dataset_id: datasetId,
  })
  return data.response
}

export async function interpret(analysisType, result) {
  const { data } = await client.post('/api/ai/interpret', {
    analysis_type: analysisType,
    result,
  })
  return data.interpretation
}

export async function recommendTest(datasetId, question) {
  const { data } = await client.post('/api/ai/recommend_test', {
    dataset_id: datasetId,
    question,
  })
  return data
}

// Dataset-level summary used by the Overview panel in DataView.
// Returns { summary, issues: [...], suggestions: [...] }.
export async function getOverview(datasetId) {
  const { data } = await client.get('/api/ai/overview', {
    params: { dataset_id: datasetId },
  })
  return data
}

// Module-aware "what should I do next" panel. module is one of
// 'data' | 'clean' | 'expand' | 'stats' | 'tests' | 'model'.
// Returns { suggestions: [{ title, description, module, action?, params? }] }.
export async function getSuggestions(datasetId, module) {
  const { data } = await client.get('/api/ai/suggestions', {
    params: { dataset_id: datasetId, module },
  })
  return data
}

// AI-generated reasoning for a single Timeline step. Cached server-side
// so repeat calls are free. Pass force=true to regenerate.
export async function explainStep(stepId, force = false) {
  const { data } = await client.post('/api/ai/explain-step', {
    step_id: stepId,
    force,
  })
  return data  // { reasoning, cached }
}

// Ask Claude to pick a snake_case name for a derived column.
export async function suggestColumnName(sourceColumns, operation, description = null) {
  const { data } = await client.post('/api/ai/suggest-column-name', {
    source_columns: sourceColumns,
    operation,
    description,
  })
  return data.name
}

// Ask Claude to pick model features given a target column.
// Returns { features: [...], reasoning }.
export async function suggestModelFeatures(datasetId, target) {
  const { data } = await client.post('/api/ai/suggest-model-features', {
    dataset_id: datasetId,
    target,
  })
  return data
}

File: backend/routes/ai.py
# ============================================================================
# routes/ai.py
# ============================================================================
# All AI-powered endpoints. Delegates to ai_service.py which talks to Claude.
# ============================================================================

from flask import Blueprint, request, jsonify
from flask_login import login_required, current_user

from database import db
from models import Dataset, Step
from services import ai_service, data_service

bp = Blueprint("ai", __name__)


@bp.route("/chat", methods=["POST"])
@login_required
def chat():
    """
    POST /api/ai/chat
    Body: {
        "message": "What tests should I run?",
        "dataset_id": 42,  (optional — gives Claude context about the data)
        "history": [ {"role": "user", "content": "..."}, ... ]  (optional)
    }

    Returns Claude's text response.
    """
    data = request.json or {}
    message = data.get("message", "").strip()
    dataset_id = data.get("dataset_id")
    history = data.get("history", [])

    if not message:
        return jsonify({"error": "Message is required"}), 400

    # --- Get dataset context if provided ---
    # This lets Claude see column names, types, and a data sample —
    # so it can answer questions like "what test should I run on age vs survived?"
    dataset_context = None
    if dataset_id:
        dataset = Dataset.query.get(dataset_id)
        if dataset and dataset.project.user_id == current_user.id:
            dataset_context = dataset

    response_text = ai_service.chat(message, history=history, dataset=dataset_context)
    return jsonify({"response": response_text})


@bp.route("/interpret", methods=["POST"])
@login_required
def interpret():
    """
    POST /api/ai/interpret
    Body: {
        "analysis_type": "t_test",
        "result": { ...raw result dict... }
    }

    Returns plain-English interpretation. Fires automatically after any test.
    """
    data = request.json or {}
    analysis_type = data.get("analysis_type", "")
    result = data.get("result", {})

    if not analysis_type or not result:
        return jsonify({"error": "analysis_type and result are required"}), 400

    text = ai_service.interpret(analysis_type, result)
    return jsonify({"interpretation": text})


@bp.route("/overview", methods=["GET"])
@login_required
def overview():
    """
    GET /api/ai/overview?dataset_id=42
    AI-generated summary of dataset health + suggested next steps.
    Uses the per-column profile as input so the prompt stays small.
    """
    dataset_id = request.args.get("dataset_id")
    if not dataset_id:
        return jsonify({"error": "dataset_id is required"}), 400

    dataset = Dataset.query.get(dataset_id)
    if not dataset or dataset.project.user_id != current_user.id:
        return jsonify({"error": "Dataset not found"}), 404

    profile = data_service.build_profile(dataset)
    result = ai_service.overview(dataset, profile)
    return jsonify(result)


@bp.route("/suggestions", methods=["GET"])
@login_required
def ai_suggestions():
    """
    GET /api/ai/suggestions?dataset_id=42&module=clean
    Returns 2-4 contextual "what should I do next" cards for the given
    module, grounded in the dataset profile.
    """
    dataset_id = request.args.get("dataset_id")
    module = (request.args.get("module") or "data").strip()
    if not dataset_id:
        return jsonify({"error": "dataset_id is required"}), 400

    dataset = Dataset.query.get(dataset_id)
    if not dataset or dataset.project.user_id != current_user.id:
        return jsonify({"error": "Dataset not found"}), 404

    profile = data_service.build_profile(dataset)
    result = ai_service.suggestions(dataset, profile, module)
    return jsonify(result)


@bp.route("/explain-step", methods=["POST"])
@login_required
def explain_step():
    """
    POST /api/ai/explain-step
    Body: { "step_id": 42, "force": false }

    Returns {"reasoning": "..."} and caches it on the Step row so the
    next caller gets an instant response. Pass force=true to ignore
    the cache and regenerate.
    """
    data = request.json or {}
    step_id = data.get("step_id")
    force = bool(data.get("force", False))
    if not step_id:
        return jsonify({"error": "step_id is required"}), 400

    step = Step.query.get(step_id)
    if not step or step.project.user_id != current_user.id:
        return jsonify({"error": "Step not found"}), 404

    if step.reasoning and not force:
        return jsonify({"reasoning": step.reasoning, "cached": True})

    try:
        text = ai_service.explain_step(step)
    except Exception as e:
        return jsonify({"error": str(e)}), 500

    step.reasoning = text
    db.session.commit()
    return jsonify({"reasoning": text, "cached": False})


@bp.route("/suggest-column-name", methods=["POST"])
@login_required
def suggest_column_name():
    """
    POST /api/ai/suggest-column-name
    Body: { "source_columns": ["age"], "operation": "zscore", "description": "z-score of age" }
    """
    data = request.json or {}
    source = data.get("source_columns") or []
    operation = (data.get("operation") or "").strip()
    if not source or not operation:
        return jsonify({"error": "source_columns and operation are required"}), 400

    try:
        name = ai_service.suggest_column_name(source, operation, data.get("description"))
    except Exception as e:
        return jsonify({"error": str(e)}), 500
    return jsonify({"name": name})


@bp.route("/suggest-model-features", methods=["POST"])
@login_required
def suggest_model_features():
    """
    POST /api/ai/suggest-model-features
    Body: { "dataset_id": 42, "target": "survived" }
    """
    data = request.json or {}
    dataset_id = data.get("dataset_id")
    target = (data.get("target") or "").strip()
    if not dataset_id or not target:
        return jsonify({"error": "dataset_id and target are required"}), 400

    dataset = Dataset.query.get(dataset_id)
    if not dataset or dataset.project.user_id != current_user.id:
        return jsonify({"error": "Dataset not found"}), 404

    profile = data_service.build_profile(dataset)
    return jsonify(ai_service.suggest_model_features(dataset, profile, target))


@bp.route("/recommend_test", methods=["POST"])
@login_required
def recommend_test():
    """
    POST /api/ai/recommend_test
    Body: {
        "dataset_id": 42,
        "question": "Is there a difference in survival between men and women?"
    }

    Claude picks the right test and explains why.
    """
    data = request.json or {}
    dataset_id = data.get("dataset_id")
    question = data.get("question", "").strip()

    if not dataset_id or not question:
        return jsonify({"error": "dataset_id and question are required"}), 400

    dataset = Dataset.query.get(dataset_id)
    if not dataset or dataset.project.user_id != current_user.id:
        return jsonify({"error": "Dataset not found"}), 404

    result = ai_service.recommend_test(dataset, question)
    return jsonify(result)
